core/embed/rust/src/ui/model_tt/component/keyboard/pin.rs

```diff
@@ -86,36 +123,50 @@ impl PinKeyboard {
     }
 }
 
-impl Component for PinKeyboard {
+impl<T> Component for PinKeyboard<T>
+where
+    T: Deref<Target = [u8]>,
+{
     type Msg = PinKeyboardMsg;
 
     fn place(&mut self, bounds: Rect) -> Rect {
         // Prompts and PIN dots display.
-        let grid = if self.minor_prompt.text().is_empty() {
-            // Make the major prompt bigger if the minor one is empty.
-            Grid::new(bounds, 5, 1)
-        } else {
-            Grid::new(bounds, 6, 1)
-        };
-        self.major_prompt.place(grid.row_col(0, 0));
-        self.minor_prompt.place(grid.row_col(0, 1));
-        self.dots.place(grid.row_col(0, 0));
+        let (header, keypad) = bounds
+            .inset(theme::borders())
+            .split_top(Self::HEADER_HEIGHT + Self::HEADER_PADDING_BOTTOM);
+        let header = header.inset(Insets::new(
+            0,
+            Self::HEADER_PADDING_SIDE,
+            Self::HEADER_PADDING_BOTTOM,
+            Self::HEADER_PADDING_SIDE,
+        ));
+        let major_area = header.translate(Self::MAJOR_OFF);
+        let minor_area = header.translate(Self::MINOR_OFF);
 
         // Control buttons.
-        let grid = Grid::new(bounds, 5, 3);
-        self.reset_btn.place(grid.row_col(4, 0));
-        self.cancel_btn.place(grid.row_col(4, 0));
-        self.confirm_btn.place(grid.row_col(4, 2));
+        let grid = Grid::new(keypad, 4, 3).with_spacing(theme::KEYBOARD_SPACING);
+
+        // Prompts and PIN dots display.
+        self.dots.place(header);
+        self.major_prompt.place(major_area);
+        self.minor_prompt.place(minor_area);
+        self.major_warning.as_mut().map(|c| c.place(major_area));
+
+        // Control buttons.
+        let reset_cancel_area = grid.row_col(3, 0);
+        self.reset_btn.place(reset_cancel_area);
+        self.cancel_btn.place(reset_cancel_area);
+        self.confirm_btn.place(grid.row_col(3, 2));
 
         // Digit buttons.
         for (i, btn) in self.digit_btns.iter_mut().enumerate() {
             // Assign the digits to buttons on a 5x3 grid, starting from the second row.
             let area = grid.cell(if i < 9 {
                 // The grid has 3 columns, and we skip the first row.
```

Comment:
"and we skip first row" is no longer true

Reply:
fixed in [e579140](https://github.com/trezor/trezor-firmware/pull/2154/commits/e5791405c4c1cb4f1738cfbffbb652f6cd691394)